File: frontend/src/types/asset.ts
export interface Asset {
  id: number;
  name: string;
  description: string;
  category: string;
  serialNumber: string;
  purchaseDate: string;
  purchasePrice: number;
  status: "active" | "inactive" | "maintenance" | "disposed";
  createdAt: string;
  updatedAt: string;
}

export interface CreateAssetRequest {
  name: string;
  description: string;
  category: string;
  serialNumber: string;
  purchaseDate: string;
  purchasePrice: number;
  status: "active" | "inactive" | "maintenance" | "disposed";
}

export interface UpdateAssetRequest {
  name: string;
  description: string;
  category: string;
  serialNumber: string;
  purchaseDate: string;
  purchasePrice: number;
  status: "active" | "inactive" | "maintenance" | "disposed";
}

export interface AssetListResponse {
  assets: Asset[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export interface AssetFilters {
  search?: string;
  category?: string;
  status?: string;
  page?: number;
  pageSize?: number;
}

export type AssetStatus = "active" | "inactive" | "maintenance" | "disposed";

// Utility types for better type safety
export interface AssetFormData
  extends Omit<CreateAssetRequest, "purchaseDate"> {
  purchaseDate: Date | string;
}

export interface AssetUpdateFormData
  extends Omit<UpdateAssetRequest, "purchaseDate"> {
  purchaseDate: Date | string;
}

// Search and filter types
export interface AssetSearchParams {
  query?: string;
  category?: string;
  status?: AssetStatus;
}

export interface PaginationParams {
  page?: number;
  pageSize?: number;
}

export interface AssetQueryParams extends AssetSearchParams, PaginationParams {}

// Response wrapper types
export interface AssetResponse {
  asset: Asset;
}

export interface CategoriesResponse {
  categories: string[];
}

File: backend/app/schemas/asset.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional
from datetime import date, datetime
from decimal import Decimal
from ..models.asset import AssetStatus

class AssetBase(BaseModel):
  name: str = Field(..., min_length=1, max_length=255)
  description: Optional[str] = None
  category: Optional[str] = Field(None, min_length=1, max_length=100)
  serial_number: str = Field(..., min_length=1, max_length=100)
  purchase_date: date
  purchase_price: Decimal = Field(..., gt=0)
  status: AssetStatus

class AssetCreate(AssetBase):
  pass

class AssetUpdate(BaseModel):
  name: Optional[str] = Field(None, min_length=1, max_length=255)
  description: Optional[str] = None
  category: Optional[str] = Field(None, min_length=1, max_length=100)
  serial_number: Optional[str] = Field(None, min_length=1, max_length=100)
  purchase_date: Optional[date] = None
  purchase_price: Optional[Decimal] = Field(None, gt=0)
  status: Optional[AssetStatus] = None

class AssetResponse(AssetBase):
  model_config = ConfigDict(from_attributes=True)
  
  id: int
  created_at: datetime
  updated_at: datetime

class AssetListResponse(BaseModel):
  assets: list[AssetResponse]
  total: int
  page: int
  page_size: int
  total_pages: int